### PR TITLE
user-guide: add troubleshooting entry for dvc-exp in shallow clones

### PR DESCRIPTION
**Pull request by pmrowla**
Closes https://github.com/iterative/dvc.org/issues/3416 (via this https://github.com/iterative/dvc.org/issues/3416#issuecomment-1089866190).

In review app: https://dvcorg-troubleshootingshallow.gtsb.io/doc/user-guide/troubleshooting#git-shallow

### CODE REVIEW

**Comment by pmrowla** [comment]
This PR adds a troubleshooting entry with specific instructions for resolving the problem in CML, github actions, and gitlab CI/CD. It does not touch any of the command reference docs suggested in https://github.com/iterative/dvc.org/issues/3416, but IMO this should be sufficient to close the original issue, given that this is an edge case that is mostly specific to CI.

**Comment by dberenbaum** [comment]
Is this intended to be in conjunction with https://github.com/iterative/dvc/pull/7791?

**Comment by pmrowla** [comment]
Yes, this can probably be merged without waiting for the DVC PR though

File: content/docs/user-guide/troubleshooting.md
Comment (dberenbaum) [inline content/docs/user-guide/troubleshooting.md:131]: Maybe include a link to https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthltdepthgt or elsewhere to explain shallow clones?

File: content/docs/user-guide/troubleshooting.md
Comment (dberenbaum) [inline content/docs/user-guide/troubleshooting.md:132]: Not sure everyone will know what we mean by "CI environments" or even what CI is. Maybe at least we can mention some of those common environments, like GitHub Actions and GitLab CI?

File: content/docs/user-guide/troubleshooting.md
Comment (dberenbaum) [inline content/docs/user-guide/troubleshooting.md:148]: Can we further clarify that users should either do the CML option or one of the GH/GL options?

**Comment by jorgeorpinel** [comment]
> IMO this should be sufficient to close the original issue, given that this is an edge case that is mostly specific to CI.

Yes, if we're going with the error handling + troubleshooting section then there's probably no need to update other docs. Wouldn't hurt either but it's kind of a deep implementation detail.

File: content/docs/user-guide/troubleshooting.md
Comment (jorgeorpinel) [inline content/docs/user-guide/troubleshooting.md:140]:
```suggestion
`dvc exp` commands use internal Git operations which may not work properly in
[shallow clones](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthltdepthgt).
Local Git repositories can be unshallowed with `git fetch --unshallow`.

Typically, however, this occurs in transient remote environments such as
Continuous Integration jobs, which use shallow clones by default. In those cases,
change their configuration to avoid shallow cloning. Common examples:
```

File: content/docs/user-guide/troubleshooting.md
Comment (jorgeorpinel) [inline content/docs/user-guide/troubleshooting.md:163]:
````suggestion
<toggle>
<tab title="CML">

When using [CML](https://cml.dev/doc), repositories can be unshallowed by using:

```cli
$ cml ci --unshallow
```

</tab>
<tab title="GitHub Actions">

Set `fetch-depth` to `0` in the `actions/checkout` action:

```yaml
- uses: actions/checkout@v3
  with:
    fetch-depth: 0
```

> More info.
> [here](https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches).

</tab>
<tab title="GitLab CI/CD">

Set the `GIT_DEPTH` env var to `0`:

```yaml
variables:
  GIT_DEPTH: "0"
```

> More info.
> [here](https://docs.gitlab.com/ee/ci/large_repositories/#shallow-cloning).

</tab>
</toggle>
````

**Review by jorgeorpinel — COMMENTED** [review]
Thanks for the content, all the info is great. I'm rephrasing and using tabs to clarify the alternative fixes (committing).

File: content/docs/user-guide/troubleshooting.md
Comment (jorgeorpinel) [reply content/docs/user-guide/troubleshooting.md:148]: Same.

File: content/docs/user-guide/troubleshooting.md
Comment (jorgeorpinel) [inline content/docs/user-guide/troubleshooting.md:146]: This ain't working on this repo yet...

https://dvcorg-troubleshootingshallow.gtsb.io/doc/user-guide/troubleshooting?tab=CML#git-shallow

````suggestion
```dvc
$ cml ci --unshallow
```
````

File: content/docs/user-guide/troubleshooting.md
Comment (jorgeorpinel) [inline content/docs/user-guide/troubleshooting.md:142]: Copy edit
```suggestion
[CML](https://cml.dev/doc) has a convenient `--unshallow` option for it's
[`ci`](https://cml.dev/doc/ref/ci) command:
```

**Comment by jorgeorpinel** [comment]
Mind reviewing when you can @dberenbaum ? Need an approval (besides mine since I took it over). Thanks

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [inline content/docs/user-guide/troubleshooting.md:143]:
```suggestion
[CML](https://cml.dev) has a convenient `--unshallow` option for its
[`ci`](https://cml.dev/doc/ref/ci) command:
```

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [inline content/docs/user-guide/troubleshooting.md:136]:
```suggestion
Continuous Integration (CI) jobs, which use shallow clones by default. In those
```

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [inline content/docs/user-guide/troubleshooting.md:135]:
```suggestion
This often occurs in transient remote environments such as
```

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [inline content/docs/user-guide/troubleshooting.md:131]: maybe hyperlink to more info (presumably some section on reflog things?)

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [inline content/docs/user-guide/troubleshooting.md:161]: is this meant to be a callout?

```suggestion
<admon type="info">

See the [GitHub Actions docs](https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches) for more information.

</admon>
```

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [inline content/docs/user-guide/troubleshooting.md:174]: is this meant to be a callout?

```suggestion
<admon type="info">

See the [GitLab CI/CD docs](https://docs.gitlab.com/ee/ci/large_repositories/#shallow-cloning) for more information.

</admon>
```

**Review by casperdcl — CHANGES_REQUESTED** [review]

File: content/docs/user-guide/troubleshooting.md
Comment (dberenbaum) [reply content/docs/user-guide/troubleshooting.md:131]: Could use https://dvc.org/doc/user-guide/experiment-management/experiments-overview#how-does-dvc-track-experiments or https://dvc.org/blog/experiment-refs.

**Review by dberenbaum — APPROVED** [review]
@jorgeorpinel I'm fine with it as is, but please take a look at the suggestions from @casperdcl before merging 🙏 .

**Comment by dberenbaum** [comment]
It looks like https://github.com/iterative/dvc/issues/6154#issuecomment-1140457383 may be a similar issue, in which case maybe we should broaden this entry beyond `dvc exp` scenarios?

File: content/docs/user-guide/troubleshooting.md
Comment (casperdcl) [reply content/docs/user-guide/troubleshooting.md:147]: resolving since `cli` should work now afaik

**Comment by dberenbaum** [comment]
> It looks like [iterative/dvc#6154 (comment)](https://github.com/iterative/dvc/issues/6154#issuecomment-1140457383) may be a similar issue, in which case maybe we should broaden this entry beyond `dvc exp` scenarios?

Turns out this was unrelated.

@jorgeorpinel I think this is ready to be merged once the comments from @casperdcl are resolved.

File: content/docs/user-guide/troubleshooting.md
Comment (jorgeorpinel) [reply content/docs/user-guide/troubleshooting.md:131]: We don't explain internal git merges anywhere, not even in Peter's blog post. Too deep

**Comment by jorgeorpinel** [comment]
comments*

**Pull request merged**
